### ISSUE TITLE
R regression example does not work

### ISSUE THREAD

**Issue by zachmayer** [dmlc/mxnet#544]
In a fresh R session:
```{R}
library(mxnet)
data(BostonHousing, package="mlbench")
train.ind = seq(1, 506, 3)
train.x = data.matrix(BostonHousing[train.ind, -14])
train.y = BostonHousing[train.ind, 14]
test.x = data.matrix(BostonHousing[-train.ind, -14])
test.y = BostonHousing[-train.ind, 14]
data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)
mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
```

Yields:
```{R}
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[15:07:01] ./dmlc-core/include/dmlc/logging.h:208: [15:07:01] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch
 Hide Traceback
 
 Rerun with Debug
 Error: [15:07:01] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch 
5 stop(structure(list(message = "[15:07:01] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch", 
    call = NULL, cppstack = NULL), .Names = c("message", "call", 
"cppstack"), class = c("Rcpp::exception", "C++Error", "error", 
"condition"))) 
4 exec$update.arg.arrays(arg.arrays, match.name, skip.null) 
3 mx.exec.update.arg.arrays(train.execs[[i]], s, match.name = TRUE) 
2 mx.model.train(symbol, ctx, input.shape, params$arg.params, params$aux.params, 
    1, num.round, optimizer = optimizer, train.data = X, eval.data = eval.data, 
    metric = eval.metric, epoch.end.callback = epoch.end.callback, 
    batch.end.callback = batch.end.callback, kvstore = kvstore) 
1 mx.model.FeedForward.create(lro, X = train.x, y = train.y, ctx = mx.cpu(), 
    num.round = 50, array.batch.size = 20, learning.rate = 2e-06, 
    momentum = 0.9, eval.metric = mx.metric.rmse) 
```

Their is either a bug in the code or with my installation.  The sonar classification example and MNIST examples work fine.  I followed the install guide here:
https://medium.com/@lelayf/installing-mxnet-for-r-on-yosemite-291470055980

```{R}
> sessionInfo()
R version 3.2.2 (2015-08-14)
Platform: x86_64-apple-darwin14.5.0 (64-bit)
Running under: OS X 10.10.5 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mxnet_0.5

loaded via a namespace (and not attached):
 [1] htmlwidgets_0.5  DiagrammeR_0.8.1 magrittr_1.5     visNetwork_0.1.2 htmltools_0.2.6  tools_3.2.2     
 [7] rstudioapi_0.3.1 Rcpp_0.12.1      codetools_0.2-14 stringi_1.0-1    data.table_1.9.6 jsonlite_0.9.17 
[13] stringr_1.0.0    digest_0.6.8     chron_2.3-47    
```

**Comment by zachmayer** [comment]
This also fails:
```{R}
mx.set.seed(0)
model <- mx.mlp(
  train.x, train.y, hidden_node=10, out_node=2, out_activation="rmse",
  num.round=20, array.batch.size=15, learning.rate=0.07, momentum=0.9,
  eval.metric=mx.metric.accuracy)
```
```{R}
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[15:10:12] ./dmlc-core/include/dmlc/logging.h:208: [15:10:12] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch
 Hide Traceback
 
 Rerun with Debug
 Error: [15:10:12] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch 
6 stop(structure(list(message = "[15:10:12] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch", 
    call = NULL, cppstack = NULL), .Names = c("message", "call", 
"cppstack"), class = c("Rcpp::exception", "C++Error", "error", 
"condition"))) 
5 exec$update.arg.arrays(arg.arrays, match.name, skip.null) 
4 mx.exec.update.arg.arrays(train.execs[[i]], s, match.name = TRUE) 
3 mx.model.train(symbol, ctx, input.shape, params$arg.params, params$aux.params, 
    1, num.round, optimizer = optimizer, train.data = X, eval.data = eval.data, 
    metric = eval.metric, epoch.end.callback = epoch.end.callback, 
    batch.end.callback = batch.end.callback, kvstore = kvstore) 
2 mx.model.FeedForward.create(out, X = data, y = label, ctx = device, 
    ...) 
1 mx.mlp(train.x, train.y, hidden_node = 10, out_node = 2, out_activation = "rmse", 
    num.round = 20, array.batch.size = 15, learning.rate = 0.07, 
    momentum = 0.9, eval.metric = mx.metric.accuracy) 
```

**Comment by thirdwing** [comment]
Hi, @zachmayer 
:+1:  for the reproducible case and traceback info. 

I have confirmed the bug in https://github.com/dmlc/mxnet/blob/master/R-package/R/model.R#L150-L154

I will fix this soon.

**Comment by tqchen** [comment]
Could due to shape convention requirement in the regression output operator not match the shape of label in the arrayIter

**Comment by thirdwing** [comment]
Not enough time today. The error is https://github.com/dmlc/mxnet/blob/master/R-package/src/executor.cc#L40-L41

The dimension is not the same.

**Comment by zachmayer** [comment]
Thanks for getting on this so fast!

**Comment by tqchen** [comment]
Should be fixed by #550

**Comment by zachmayer** [comment]
I'll rebuild and check!

**Comment by zachmayer** [comment]
I still get the same issue, even after re-pulling the repo, building the R package, and running `R CMD INSTALL mxnet_0.5.tar.gz`.  Looking at your PR, you changed some python code and some C++ code. I think you may also need to update the R code and the Rcpp code:
https://github.com/dmlc/mxnet/blob/master/R-package/R/model.R#L150-L154
https://github.com/dmlc/mxnet/blob/master/R-package/src/executor.cc#L40-L41

**Comment by thirdwing** [comment]
Hi,@zachmayer 

Just come back from the exam and fix this (see #560 ). Actually it is a very simple problem. We met an inconsistency of dimensions between R and the backend.

Now you should have something like below:

```r
> library(mxnet)
> data(BostonHousing, package="mlbench")
> train.ind = seq(1, 506, 3)
> train.x = data.matrix(BostonHousing[train.ind, -14])
> train.y = BostonHousing[train.ind, 14]
> test.x = data.matrix(BostonHousing[-train.ind, -14])
> test.y = BostonHousing[-train.ind, 14]
> data <- mx.symbol.Variable("data")
> fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
> lro <- mx.symbol.LinearRegressionOutput(fc1)
> mx.set.seed(0)
> model <- mx.model.FeedForward.create(
+   lro, X=train.x, y=train.y,
+   ctx=mx.cpu(), num.round=50, array.batch.size=20,
+   learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[1] Train-rmse=16.063282524034
[2] Train-rmse=12.2792375712573
[3] Train-rmse=11.1984634005885
[4] Train-rmse=10.2645236892904
......
```

Just let us know if you met new problems.

**Comment by thirdwing** [comment]
I am afraid I broke something else.

**Comment by zachmayer** [comment]
works now.  Thank you!

**Comment by kimquy06** [comment]
I've just installed and got the same error when trying regression.

**Comment by thirdwing** [comment]
@kimquy06 Sorry I can't reproduce this error any more.

Can you tell me your OS and R version? If you are on Linux/Mac, the compiler version will be also useful.

**Comment by kimquy06** [comment]
Hi @thirdwing , I use R version 3.2.2, gcc (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4

**Comment by thirdwing** [comment]
Then it is quite weird, since I can run the code without error?

Did you checkout the latest code and compile it by yourself?

**Comment by kimquy06** [comment]
Yes, I did twice yesterday

**Comment by zhugejun** [comment]
I tried the example above, but I got the following error message:

```
> model <- mx.model.FeedForward.create(lro, X=train.x, y=train.y,
+                                      ctx=mx.cpu(), num.round=50, array.batch.size=20,
+                                      learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse)
Auto detect layout of input matrix, use rowmajor..
Start training with 1 devices
[14:20:18] ./dmlc-core/include/dmlc/logging.h:208: [14:20:18] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch
Error: [14:20:18] src/ndarray/ndarray.cc:159: Check failed: from.shape() == to->shape() operands shape mismatch
```
How can I fix it? I am using Ubuntu 14.04, R 3.2.3. Thanks.

**Comment by thirdwing** [comment]
I am really sorry I can't reproduce your errors.

The code runs well on my machine and we test all examples on Travis CI.

My suggestion is following:

(1) checkout the latest version of code;
(2) `make clean_all`
(3) `make`
(4) `make rpkg; R CMD INSTALL mxnet_0.5.tar.gz`

If you still have errors, please give me more info about your machine, including whether you are using GPU, 64 or 32 bit machine, and your R `sessionInfo()`.

**Comment by zhugejun** [comment]
Thank you for your quick reply.

I still got some error when running the codes you provided.

```
Error : .onLoad failed in loadNamespace() for 'mxnet', details:
  call: dyn.load(file, DLLpath = DLLpath, ...)
  error: unable to load shared object '/home/gejun/R/x86_64-pc-linux-gnu-library/3.2/mxnet/libs/libmxnet.so':
  libcudart.so.7.5: cannot open shared object file: No such file or directory
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/gejun/R/x86_64-pc-linux-gnu-library/3.2/mxnet’
```

I am using Ubuntu 14.04 64-bit. Thinkpad w550s with Nvidia K620M. 
```
> sessionInfo()
R version 3.2.3 (2015-12-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 14.04.3 LTS

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mxnet_0.5     mlbench_2.1-1

loaded via a namespace (and not attached):
 [1] htmlwidgets_0.5  DiagrammeR_0.8.1 magrittr_1.5     visNetwork_0.2.0 htmltools_0.3    tools_3.2.3      rstudioapi_0.4.0
 [8] Rcpp_0.12.2      codetools_0.2-14 stringi_1.0-1    data.table_1.9.6 jsonlite_0.9.19  stringr_1.0.0    digest_0.6.9    
[15] chron_2.3-47  
```
Thank you.

**Comment by thirdwing** [comment]
Did you see the errors when installing, or loading the package?

Are you using Rstudio, or other IDEs? If so, please try again in the R console directly.

**Comment by vc273** [comment]
the example works when num_hidden = 1 for fc1, but fails with the original error, as soon as num_hidden > 1. i am using
```
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.11.6 (El Capitan)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] mxnet_0.7        data.table_1.9.6

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.7      codetools_0.2-14 visNetwork_1.0.1 digest_0.6.10    chron_2.3-47     plyr_1.8.4       jsonlite_1.1     magrittr_1.5     scales_0.4.0     stringi_1.1.1    rstudioapi_0.6   DiagrammeR_0.8.4 tools_3.3.1      stringr_1.1\
.0
[15] htmlwidgets_0.7  munsell_0.4.3    igraph_1.0.1     influenceR_0.1.0 colorspace_1.2-6 htmltools_0.3.5
```
